add rule delimiter option

ahumphreys87 opened the pull request:
@tonyganch as requested opened #404 against dev.

tonyganch commented:
I'd like to discuss some things first, so please don't rush to change anything in your code :)

As I understand, you've made this option as a modified copy of `space-between-declarations`. But `space-between-declarations` works differently than what you propose here: you can set any string as a _declaration_ delimiter (even an empty one), but you can set only newlines as a _rule_ delimiter.
For example, there is no way to reflect this code style with your implementation:
```scss
.a { ... } .b { ... }
```

Personally, I find this confusing and think that all options should work in the same way, so there should be a way to set one whitespace as a rule delimiter. But I guess all the people who leave comments in #210 (read: all those who really need this feature) think otherwise. It bothers me but I'm ok with it as long as you're happy. Just wanna make sure that you understand: your implementation of #210 will be the only option where number will mean number of _newlines_ while everywhere else it's number of _whitespaces_. Do you still want this option to accept only newlines?

tonyganch commented:
I'd like to propose something based on your decision:
(1) if you want the option to accept whitespaces as well, let's rename it to `space-between-rulesets` and fix the code;
(2) if you want the option to accept only newlines, let's rename it to `lines-between-rulesets` and make it accept only number, not string.
What do you think?

ahumphreys87 commented:
I see where you are coming from. I implemented this based on the
conversation from the thread - one of my colleagues @alexnewmann
contributed to that conversation.

I can understand your concerns here, would it be better if we removed the
Number option? And stayed with strings? I think a ' ' would cater for the
use case you describe above?

Or would that also stray from the other options?

On Sunday, 5 July 2015, Tony Ganch <notifications@github.com> wrote:

> I'd like to discuss some things first, so please don't rush to change
> anything in your code :)
>
> As I understand, you've made this option as a modified copy of
> space-between-declarations. But space-between-declarations works
> differently than what you propose here: you can set any string as a
> *declaration* delimiter (even an empty one), but you can set only
> newlines as a *rule* delimiter.
> For example, there is no way to reflect this code style with your
> implementation:
>
> .a { ... } .b { ... }
>
> Personally, I find this confusing and think that all options should work
> in the same way, so there should be a way to set one whitespace as a rule
> delimiter. But I guess all the people who leave comments in #210
> <https://github.com/csscomb/csscomb.js/issues/210> (read: all those who
> really need this feature) think otherwise. It bothers me but I'm ok with it
> as long as you're happy. Just wanna make sure that you understand: your
> implementation of #210 <https://github.com/csscomb/csscomb.js/issues/210>
> will be the only option where number will mean number of *newlines* while
> everywhere else it's number of *whitespaces*. Do you still want this
> option to accept only newlines?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/csscomb/csscomb.js/pull/405#issuecomment-118619917>.
>

ahumphreys87 commented:
Sorry I didn't see this comment before I posted. (2) sounds good for ease
of getting this option available to those who need it. I think this would
be my preference and it certainly serves my use case

On Sunday, 5 July 2015, Tony Ganch <notifications@github.com> wrote:

> I'd like to propose something based on your decision:
> (1) if you want the option to accept whitespaces as well, let's rename it
> to space-between-rulesets and fix the code;
> (2) if you want the option to accept only newlines, let's rename it to
> lines-between-rulesets and make it accept only number, not string.
> What do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/csscomb/csscomb.js/pull/405#issuecomment-118621944>.
>

tonyganch commented on src/options/rule-delimiter.js:
```diff
@@ -0,0 +1,106 @@
+
+        process: function(ast, config) {
+            valueFromSettings = this.value;
+            blockIndent = config['block-indent'];
```
Ok, so `lines-between-rulesets`!
`block-indent` option is useless here because `lines-between-rulesets` should be independent of other options (user can have no `block-indent` in his config). So we need to remove this line.

tonyganch commented on src/options/rule-delimiter.js:
```diff
@@ -0,0 +1,106 @@
+            if (value.indexOf('\n') > -1) {
+                // TODO: Check that it works for '' block indent value <tg>
+                if (blockIndent) {
+                    value += new Array(level).join(blockIndent);
```
Here you use `blockIndent` if it's set in config, but it may not be there.

tonyganch commented on src/options/rule-delimiter.js:
```diff
@@ -0,0 +1,106 @@
+            var prevNode = node.get(index - 1);
+            if (prevNode) {
+                if (prevNode.is('space')) {
+                    prevNode.content = value;
```
I see two cases here.
```scss
// Example (1)
.a {...} .b {...}

// Example (2)
.a {...}
   .b {...}
```

In case (1) you can just replace `prevNode.content` with `value` as you do now.
In case (2) you can replace all newlines in `prevNode.content` with `value`.
Something like: "let `y` be the index of the last newline, then replace content with `value` + substring from `y` to the end of space node". This should preserve indent without using `block-indent`.

Please let me know if I'm understandable at all :)

ahumphreys87 replied:
yup got it! im make these changes now

On Mon, Jul 6, 2015 at 11:34 AM, Tony Ganch <notifications@github.com>
wrote:

> In src/options/rule-delimiter.js
> <https://github.com/csscomb/csscomb.js/pull/405#discussion_r33922422>:
>
> > +                } else {
> > +                    node.insert(index, space);
> > +                }
> > +            }
> > +        });
> > +    }
> > +
> > +    function findRuleSets(node) {
> > +        var space = gonzales.createNode({type: 'space', content: value});
> > +
> > +        node.forEach('ruleset', function(ruleSetNode, index) {
> > +            // For every ruleset - check for preceding space
> > +            var prevNode = node.get(index - 1);
> > +            if (prevNode) {
> > +                if (prevNode.is('space')) {
> > +                    prevNode.content = value;
>
> I see two cases here.
>
> // Example (1).a {...} .b {...}
> // Example (2).a {...}
>    .b {...}
>
> In case (1) you can just replace prevNode.content with value as you do
> now.
> In case (2) you can replace all newlines in prevNode.content with value.
> Something like: "let y be the index of the last newline, then replace
> content with value + substring from y to the end of space node". This
> should preserve indent without using block-indent.
>
> Please let me know if I'm understandable at all :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/csscomb/csscomb.js/pull/405/files#r33922422>.
>

ahumphreys87 commented:
@tonyganch updated now. Tests still seem to be failing though. Mainly due to spacing, and the empty block. Think I know the issue with the `mixin` thats handled by `sort-order` so thats independent, ill updated that one.

ahumphreys87 commented:
@tonyganch I think i know what the issue is with the tests. The space node doesn't actually contain the "spaces" for the indentation - any ideas?

tonyganch commented:
@ahumphreys87, can you show me any example with code and parsing tree?

ahumphreys87 commented:
ok apologies in advance for the formatting:

```
{ type: 'space',
  content: '\n\n',
  syntax: 'less',
  start: { line: 3, column: 19 },
  end: { line: 4, column: 2 } }
{ type: 'ruleset',
  content:
   [ { type: 'selector',
       content: [Object],
       syntax: 'less',
       start: [Object],
       end: [Object] },
     { type: 'block',
       content: [Object],
       syntax: 'less',
       start: [Object],
       end: [Object] } ],
  syntax: 'less',
  start: { line: 4, column: 3 },
  end: { line: 8, column: 3 } }
{ type: 'selector',
  content:
   [ { type: 'simpleSelector',
       content: [Object],
       syntax: 'less',
       start: [Object],
       end: [Object] } ],
  syntax: 'less',
  start: { line: 4, column: 3 },
  end: { line: 4, column: 7 } }
{ type: 'simpleSelector',
  content:
   [ { type: 'class',
       content: [Object],
       syntax: 'less',
       start: [Object],
       end: [Object] },
     { type: 'space',
       content: ' ',
       syntax: 'less',
       start: [Object],
       end: [Object] } ],
  syntax: 'less',
  start: { line: 4, column: 3 },
  end: { line: 4, column: 7 } }
{ type: 'class',
  content:
   [ { type: 'ident',
       content: 'bar',
       syntax: 'less',
       start: [Object],
       end: [Object] } ],
  syntax: 'less',
  start: { line: 4, column: 3 },
  end: { line: 4, column: 6 } }
```

ahumphreys87 commented:
this shows up when processing the following in the tests:

```
.foo {
  .border-radius(5px);
  background: red;
  .bar {
    .test {
      height: 50px;
    }
  }
}
```

ahumphreys87 commented:
ok ive found 1 more possible issue - the `process` block is called many times, ill try restrict this to only process `stylesheet` nodes. possibly an issue with this...

tonyganch commented:
@ahumphreys87, there are a few ast methods that may be of some help:

(1) `ast.traverse(callback)` — calls the callback function for every node in tree. This method is available for every node.
https://github.com/tonyganch/gonzales-pe/blob/dev/src/node/basic-node.js#L164

(2) `ast.traverseByType(type, callback)` — calls the callback function for every node of given type. This method is available only for root (`stylesheet` type) node.
https://github.com/tonyganch/gonzales-pe/blob/dev/src/node/root-node.js#L21

(3) `ast.traverseByTypes(types, callback)` — same as above but accepts a list of types.
https://github.com/tonyganch/gonzales-pe/blob/dev/src/node/root-node.js#L49

Those methods may be useful instead of calling `ast.forEach()` recursively.

ahumphreys87 commented:
sorry @tonyganch missed this comment! Ive got it working now and all tests passing. I've had a look into the empty block case as well: 

```
.bar {
  .test {
```

`lines-between-rulesets` will insert a space between these but this is then modified by `space-after-opening-brace` I think this is the correct way to do it and offers the most flexibility going forward. It means users for do in fact wish for:

```
.bar {

  .test {
```

can achieve this without `lines-between-rulesets` interfering.

tonyganch commented:
Sorry, I've been out of town, will comment on everything later tonight!

tonyganch commented:
@ahumphreys87, the option should not add a line before a ruleset if that ruleset is the first thing in node, because option's name is about lines _between_, not lines before. If we make `lines-before-rulesets` instead, it will make configuring csscomb for custom code style more difficult, because a user will need to think about which option overrides what.
See https://github.com/csscomb/csscomb.js/issues/210#issuecomment-51928119 and https://github.com/csscomb/csscomb.js/issues/210#issuecomment-53848501

ahumphreys87 commented:
@tonyganch sorry for the long delay on this - I've correct the issue with adding a line when the ruleset is the first in the node.

I think this is ready now - give me a shout if it needs anything else though

tonyganch commented on src/options/lines-between-rulesets.js:
```diff
@@ -0,0 +1,105 @@
+    return {
+        name: 'lines-between-rulesets',
+
+        syntax: ['scss', 'css', 'less'],
```
@ahumphreys87, do you think that `sass` does not need this option?

tonyganch commented on src/options/lines-between-rulesets.js:
```diff
@@ -0,0 +1,105 @@
+                value = new Array(Math.round(value) + 2).join('\n');
+            } else {
+                var err = 'The option only accepts numbers';
+                err += ' , you provided %s';
```
(1) There is no need for a space before a comma.
(2) `%s` will not expand in this case.
Here is what I get, for example:
```
> plugin.setValue('foo')
Error: The option only accepts numbers , you provided %s
```
I think the first line ("only accepts numbers") is enough.

tonyganch commented on doc/options.md:
````diff
@@ -905,6 +906,66 @@ a
+}
+```
+
+Example: `{ "lines-between-rulesets": "\n\n" }`
````
This part can be removed since the option does not accept strings anymore

tonyganch commented:
Found a failing test:
```scss
// Before
.a{}.b{}.c{}

// After, "lines-between-rulesets": 2
.a{}


.b{}.c{}
```

tonyganch commented on doc/options.md:
```diff
@@ -29,6 +29,7 @@ Here is a full list in the same order they are applied while processing css:
 - [unitless-zero](#unitless-zero)
 - [tab-size](#tab-size)
 - [vendor-prefix-align](#vendor-prefix-align)
+- [lines-between-rulesets](#lines-between-rulesets)
```
This line should be between `always-semicolon` and `remove-empty-rulesets`

nielsenramon commented:
@ahumphreys87 @tonyganch What's left to do on this PR? Can I help? 
I would really like to see this merged :)

siriniok commented:
Guys, please, very needed feature.

ahumphreys87 commented:
Sorry guys, been really busy recently. Let me get back on this

On Thursday, 10 September 2015, siriniok <notifications@github.com> wrote:

> Guys, please, very needed feature.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/csscomb/csscomb.js/pull/405#issuecomment-139294700>.
>

joaocunha commented:
+1.

In the meanwhile, is disabling line breaks handling entirely something hard to achieve? That would help a lot as an initial step.

TomNeyland-FD commented:
+1

TomNeyland-FD commented:
@tonyganch any status update?

ahumphreys87 commented:
This one is on me guys, sorry I just haven't had time to look at it. It is
very close though, @tonyganch just found some test cases that won't pass
currently. Fancy lending a hand?

On Thursday, 29 October 2015, Thomas Neyland <notifications@github.com>
wrote:

> @tonyganch <https://github.com/tonyganch> any status update?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/csscomb/csscomb.js/pull/405#issuecomment-152283831>.
>

AoDev commented:
Hi, I've been wanting this feature for a while, and followed it for like a year or so. Is there something I can do to speed this up?

ahumphreys87 commented:
@AoDev unfortunately I havent had time to look at the comments @tonyganch left for me. I think this is pretty close though - if you would like to take it over that would be welcome

AoDev commented:
@ahumphreys87 Could you tell us what is working and what is missing? From the comments I can't really tell. Just make it pass some tests, write more tests, the expected behaviour (name of the rule, value for the rule) is definite or still need some thinking?

tonyganch commented:
@AoDev, the issue is fine in general. I need this example to pass:
```sass
// Before
.a{}.b{}.c{}

// After, as it is right now, with config {"lines-between-rulesets": 2}
.a{}


.b{}.c{}

// With the same config, as it should be:
.a{}


.b{}


.c{}
```
So I need one more test to be added for this issue and some code to be changed so it passes.

AoDev commented:
@tonyganch I have forked @ahumphreys87 branch. I am not too sure where I should send the PR but you can see WIP here:

https://github.com/AoDev/csscomb.js/pull/1

Currently:

* CSS test seems to work
* less and scss fail

tonyganch commented:
@AoDev, could you open a pr for `csscomb:dev` branch please?
